mtz-ajax-interceptor-behavior.html

```diff
@@ -1,13 +1,136 @@
+     *
+     * @param {String} eventName - the event name that triggered the interceptor
+     * @param {Event} event - the event that triggered the interceptor logic
+     */
+    runInterceptors(eventName, event) {
+      const interceptors = this.interceptors[eventName];
+      if (!event || !interceptors) return;
 
-<script>
-  window.mtz = mtz || {};
+      // Override stopImmediatePropagation so we can break out if needed.
+      const stopImmediatePropagationFnc = event.stopImmediatePropagation;
+      let stopPropagation = false;
+      event.stopImmediatePropagation = () => {
+        stopPropagation = true;
+      };
+
+      for (const interceptor of Array.from(interceptors).sort((a, b) => b.priority - a.priority)) {
+        if (stopPropagation)
+          break;
+        interceptor.intercept(event, eventName);
+      }
+
+      // Restore stopImmediatePropagation incase it's needed.
+      event.stopImmediatePropagation = stopImmediatePropagationFnc;
+
+      // If propagation was stopped, ensure that the original logic is also called.
+      if (stopPropagation)
+        event.stopImmediatePropagation();
+    },
+    /**
+     * Registers an event listener for the provided event name against the runInterceptors method.
+     * @private
+     *
+     * @param {String} eventName
+     */
+    __registerEventListener(eventName) {
+      this.addEventListener(eventName, this.runInterceptors);
+    },
+    /**
+     * Adds the interceptor stored in event.detail to each event referenced in events, if a new event group is added
+     *  then __registerEventListener is called.
+     * @private
+     *
+     * @param {CustomEvent} event
+     */
+    __registerInterceptor(event) {
+      const events = event.detail.events;
+      if (!events || !events.length) return;
+
+      event.stopPropagation();
+
+      // Element registered, now add an unregister listener
+      event.detail.addEventListener('unregister-interceptor', this.__unregisterInterceptor);
+
+      events.forEach((e) => {
+        if (!this.interceptors[e.name]) {
+          this.interceptors[e.name] = new Set();
+          this.__registerEventListener(e.name);
+        }
+        this.interceptors[e.name].add(event.detail);
+      });
+
+      this.notifyPath('interceptors', this.interceptors);
+    },
+    /**
+     * Removes the event listener for the provided event name against the runInterceptors method.
+     * @private
+     *
+     * @param {String} eventName
+     */
+    __unregisterEventListener(eventName) {
+      this.removeEventListener(eventName, this.runInterceptors);
+    },
+    /**
+     * Removes the interceptor from each event Set it's part of, if the set is emptied out then
+     *  __unregisterEventListener is called.
+     * @private
+     *
+     * @param {CustomEvent} event
+     */
+    __unregisterInterceptor(event) {
+      const events = event.detail.events;
```

Comment:
Don't have to, this is that weird case we talked about where it's detached from the DOM already so you have to grab the element reference and explicitly attach it, which means it could never bubble even if it wanted to, however if multiple elements are explicitly doing that process I don't want to block any propagation.